src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
 
 void IUtilities::updateImageViaStagingBuffer(
     IGPUQueue* queue,
-    asset::ICPUBuffer const* srcBuffer, const core::SRange<const asset::IImage::SBufferCopy>& regions, video::IGPUImage* dstImage, asset::IImage::E_LAYOUT dstImageLayout,
+    asset::ICPUBuffer const* srcBuffer, asset::E_FORMAT srcFormat, video::IGPUImage* dstImage, asset::IImage::E_LAYOUT dstImageLayout, const core::SRange<const asset::IImage::SBufferCopy>& regions,
     uint32_t waitSemaphoreCount, IGPUSemaphore* const* semaphoresToWaitBeforeOverwrite, const asset::E_PIPELINE_STAGE_FLAGS* stagesToWaitForPerSemaphore,
     const uint32_t signalSemaphoreCount, IGPUSemaphore* const* semaphoresToSignal
 )
 {
     auto fence = m_device->createFence(static_cast<IGPUFence::E_CREATE_FLAGS>(0));
-    updateImageViaStagingBuffer(fence.get(),queue,srcBuffer,regions,dstImage,dstImageLayout,waitSemaphoreCount,semaphoresToWaitBeforeOverwrite,stagesToWaitForPerSemaphore,signalSemaphoreCount,semaphoresToSignal);
+    updateImageViaStagingBuffer(fence.get(),queue,srcBuffer,srcFormat,dstImage,dstImageLayout,regions,waitSemaphoreCount,semaphoresToWaitBeforeOverwrite,stagesToWaitForPerSemaphore,signalSemaphoreCount,semaphoresToSignal);
     auto* fenceptr = fence.get();
     m_device->blockForFences(1u,&fenceptr);
 }
 
+ImageRegionIterator::ImageRegionIterator(
+    const core::SRange<const asset::IImage::SBufferCopy>& copyRegions,
+    IPhysicalDevice::SQueueFamilyProperties queueFamilyProps,
+    asset::ICPUBuffer const* srcBuffer,
+    asset::E_FORMAT srcImageFormat,
+    video::IGPUImage* const dstImage
+)
+    : regions(copyRegions)
+    , srcBuffer(srcBuffer)
+    , dstImage(dstImage)
+    , srcImageFormat(srcImageFormat)
+    , currentBlockInRow(0u)
+    , currentRowInSlice(0u)
+    , currentSliceInLayer(0u)
+    , currentLayerInRegion(0u)
+    , currentRegion(0u)
+{
+    dstImageFormat = dstImage->getCreationParameters().format;
+    if(srcImageFormat == asset::EF_UNKNOWN)
+        srcImageFormat = dstImageFormat;
+    asset::TexelBlockInfo dstImageTexelBlockInfo(dstImageFormat);
+
+    memcpy(&minImageTransferGranularity, &queueFamilyProps.minImageTransferGranularity, sizeof(VkExtent3D));
+
+    // bufferOffsetAlignment:
+        // [x] If Depth/Stencil -> must be multiple of 4
+        // [ ] If multi-planar -> bufferOffset must be a multiple of the element size of the compatible format for the aspectMask of imagesubresource
+        // [x] If Queue doesn't support GRAPHICS_BIT or COMPUTE_BIT ->  must be multiple of 4
+        // [x] bufferOffset must be a multiple of texel block size in bytes
+    bufferOffsetAlignment = dstImageTexelBlockInfo.getBlockByteSize();
+    if(asset::isDepthOrStencilFormat(dstImageFormat))
+        bufferOffsetAlignment = std::lcm(bufferOffsetAlignment, 4u);
```

Comment:
you probably want to use `max(limits.optimalBufferCopyRowPitchAlignment,optimalBufferCopyOffsetAlignment,4)` do the lcm and call it a day.